pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,231 @@
+	active  atomic.Bool
+	cb      circuitbreaker.CircuitBreaker[any]
+
+	// testRequestDelay is needed for testing purposes to simulate long lasting requests
+	testRequestDelay time.Duration
+}
+
+func newCircuitBreaker(cfg CircuitBreakerConfig, isActive bool, logger log.Logger, registerer prometheus.Registerer) *circuitBreaker {
+	active := atomic.NewBool(isActive)
+	cb := circuitBreaker{
+		cfg:    cfg,
+		logger: logger,
+		active: *active,
+	}
+
+	circuitBreakerTransitionsCounterFn := func(metrics *circuitBreakerMetrics, state circuitbreaker.State) prometheus.Counter {
+		return metrics.circuitBreakerTransitions.WithLabelValues(state.String())
+	}
+
+	cbBuilder := circuitbreaker.Builder[any]().
+		WithFailureThreshold(cfg.FailureThresholdPercentage).
+		WithDelay(cfg.CooldownPeriod).
+		OnClose(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.ClosedState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is closed", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.OpenState).Inc()
+			level.Warn(logger).Log("msg", "circuit breaker is open", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.HalfOpenState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is half-open", "previous", event.OldState, "current", event.NewState)
+		})
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThresholdPercentage)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThresholdPercentage, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.cb = cbBuilder.Build()
+	cb.metrics = newCircuitBreakerMetrics(registerer, cb.cb.State)
+	return &cb
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+	if statusCode == codes.DeadlineExceeded {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	if cb == nil {
+		return false
+	}
+	return cb.active.Load()
```

Comment:
We can make this shorter:
```suggestion
	return cb != nil && cb.active.Load()
```